test/helpers.py

```diff
@@ -89,3 +112,166 @@ def _wait_until_query_success(
             f"Query ID: {sfqid}"
             f"Final query status: {status}"
         )
+
+
+def create_pyarrow_iterator(input_data, use_table_iterator):
+    # create nanoarrow based iterator
+    return (
+        NanoarrowPyArrowRowIterator(
+            None,
+            input_data,
+            ArrowConverterContext(
+                session_parameters={"TIMEZONE": "America/Los_Angeles"}
+            ),
+            False,
+            False,
+            False,
+        )
+        if not use_table_iterator
+        else NanoarrowPyArrowTableIterator(
+            None,
+            input_data,
+            ArrowConverterContext(
+                session_parameters={"TIMEZONE": "America/Los_Angeles"}
+            ),
+            False,
+            False,
+            False,
+        )
+    )
+
+
+def create_old_pyarrow_iterator(input_data, use_table_iterator):
+    # created vendored arrow based iterator
+    iterator = PyArrowIterator(
+        None,
+        io.BytesIO(input_data),
+        ArrowConverterContext(session_parameters={"TIMEZONE": "America/Los_Angeles"}),
+        False,
+        False,
+        False,
+    )
+    if use_table_iterator:
+        iterator.init_table_unit()
+
+
+def create_arrow_iterator_method(use_nanoarrow_iterator):
+    if use_nanoarrow_iterator:
+        return create_pyarrow_iterator
+    else:
+        return create_old_pyarrow_iterator
+
+
+def _arrow_error_stream_chunk_remove_single_byte_test(
+    use_table_iterator, use_nanoarrow_iterator=True
+):
+    # this test removes single byte from the input bytes
+    with open(os.path.join(os.path.dirname(__file__), "data", "test_arrow_data")) as f:
+        b64data = f.read()
+
+    decode_bytes = base64.b64decode(b64data)
+    exception_result = []
+    succeeded_result = []
+    result_array = []
+    for i in range(len(decode_bytes)):
+        try:
+            # removing the i-th char in the bytes
+            iterator = create_arrow_iterator_method(use_nanoarrow_iterator)(
+                decode_bytes[:i] + decode_bytes[i + 1 :], use_table_iterator
+            )
+            for k in iterator:
+                result_array.append(k)
+            succeeded_result.append(i)
+        except Exception as e:
+            with pytest.raises(UnboundLocalError):
+                for _ in iterator:
+                    pass
```

Comment:
Is this needed?

Reply:
you mean the test? yes. I explicitly add the test because in one version of my previous implementation, my `__next__` is wrong and for loop for twice will get a different error other than the `UnboundLocalError`

Reply:
Curious: Why will we hit `UnboundLocalError`?

Reply:
in the line `iterator=create_arrow_iterator_method`,
`create_arrow_iterator_method` will raise error, leading to the variable `iterator` undefined, when we want to use `iterator` it will cause `UnboundLocalError`.

this is python bebahvior